pipeline/api/cloud.py

```diff
@@ -145,6 +155,103 @@ def upload_python_object_to_file(self, obj, remote_path) -> FileGet:
             io.BytesIO(python_object_to_hex(obj).encode()), remote_path
         )
 
+    def _initialise_direct_pipeline_file_upload(self, file_size: int) -> str:
+        """Initialise a direct multi-part pipeline file upload"""
+        direct_upload_schema = PipelineFileDirectUploadInitCreate(file_size=file_size)
+        response = self._post(
+            "/v2/pipeline-files/initiate-multipart-upload", direct_upload_schema.dict()
+        )
+        direct_upload_get = PipelineFileDirectUploadInitGet.parse_obj(response)
+        return direct_upload_get.pipeline_file_id
+
+    def _direct_upload_pipeline_file_chunk(
+        self, data: bytes, pipeline_file_id: str, part_num: int
+    ) -> dict:
+        """Upload a single chunk of a multi-part pipeline file upload.
+
+        Returns the metadata associated with this upload (this is needed to pass into
+        the finalisation step).
+        """
+        # get presigned URL
+        part_upload_schema = PipelineFileDirectUploadPartCreate(
+            pipeline_file_id=pipeline_file_id, part_num=part_num
+        )
+        response = self._post(
+            "/v2/pipeline-files/presigned-url", part_upload_schema.dict()
+        )
+        part_upload_get = PipelineFileDirectUploadPartGet.parse_obj(response)
+        # upload file chunk
+        # convert data to hex
+        data = data.hex().encode()
+        response = requests.put(part_upload_get.upload_url, data=data)
+        etag = response.headers["ETag"]
+        return {"ETag": etag, "PartNumber": part_num}
+
+    def _finalise_direct_pipeline_file_upload(
+        self, pipeline_file_id: str, multipart_metadata: List[dict]
+    ) -> PipelineFileGet:
+        """Finalise the direct multi-part pipeline file upload"""
+        finalise_upload_schema = PipelineFileDirectUploadFinaliseCreate(
+            pipeline_file_id=pipeline_file_id,
+            multipart_metadata=multipart_metadata,
+        )
+        response = self._post(
+            "/v2/pipeline-files/finalise-multipart-upload",
+            finalise_upload_schema.dict(),
+        )
+        return PipelineFileGet.parse_obj(response)
+
+    def upload_pipeline_file(self, pipeline_file) -> PipelineFileVariableGet:
+        """Upload PipelineFile given by pipeline_file.
```

Comment:
Can we type `pipeline_file`? Then the docstring can be a bit more to the point.

```suggestion
    def upload_pipeline_file(self, pipeline_file: PipelineFile) -> PipelineFileVariableGet:
        """Upload the pipeline input file.
```

Reply:
I wanted to but I don't think we can due to circular dependencies (i.e. if I use a type annotation I have to import it at the module level and this breaks things). Would be really good to fix these circular imports at some point

Reply:
Yeah but there's an `import` just above it in the same method

Reply:
D'oh, I'm blind. Ignore me!